### DISCUSSION TITLE
Kotlin JS Support?

### DISCUSSION THREAD

**Discussion by pusolito** [grpc/grpc-web#1371] [Q&A - General]
I haven't used gRPC Web yet, though I'm familiar with other platforms. I'd love to use it for Web, but with projects that use Kotlin JS. I know it's possible to consume JS types in a Kotlin JS project. But I'm curious if there are any plans to have Kotlin as a first class language for this as well.

I know there is a separate gRPC-kotlin project.  But it seems to focus on the JVM.

**Comment by sampajano** [comment]
@pusolito Thanks for the interest and inquiry!

> But I'm curious if there are any plans to have Kotlin as a first class language for this as well.

Currently, there is no plan for Kotlin support as you were the first to bring this under our radar :)

But if there's more clarity on what this would mean (e.g. how much code would be needed, like maybe this means something similar to a new `index.d.ts` file?), this could be considered for sure. (Maybe the community can help add it :))

**Reply by pusolito** [reply]
The current d.ts output is compatible with Kotlin, since it can generate Kotlin hooks for the JS. But this requires a separate step and the interop generation is still a moving target for the language.

Ideally, Kotlin code would be generated as an output format. That would require a reasonable amount of work since the C++ layer (which I think does the generation now) would need new logic for this line it has for TS.

**Reply by sampajano** [reply]
@pusolito Thanks for the info! Good to understand what work is required here!

We currently don't have plans to support Kotlin outputs. But as the language is picking up in popularity, we'd definitely be keeping this on our radar.

If the community wishes to provide a more native solution in any way today (e.g. by adding a new code-gen binary for Kotlin), i would be happy to assist in reviewing / merging it also. (Although i'm still a bit blurry on how that would invoke our underlying JS code.)

Thanks! :)

**Reply by pusolito** [reply]
Thanks for following up. There's a project that already supports Kotlin for gRPC-node ([protokt](https://github.com/open-toast/protokt)) and the author (@andrewparmet) is interested in understanding what it would take to [support gRPC-web](https://github.com/open-toast/protokt/issues/202#issuecomment-1756044589_) as well.

**Reply by sampajano** [reply]
Aha very Interesting! Thanks for informing us!

I'm really not familiar with the Kotlin compilation process and how it fits into what we already have..

But if there's a sufficient level of interest here i'm definitely happy to stay involved and see if there's anything we can do from our side 😃
